<#64> feat : Thumb api 작성

nrudev opened the pull request:
## 내용
- Thumb 등록 기능 구현
- Thumb 취소 기능 구현
- Thumb 여부 확인

## 참고
- #64

nrudev committed:
feat : thumb 등록 기능 구현

nrudev committed:
feat : Thumb 취소 기능 구현

nrudev committed:
feat : Thumb 여부 확인 기능 구현

nrudev committed:
Merge branch 'develop' of https://github.com/J2KB-5th/chippo into feature/64-back-thumbApi

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/controller/ThumbController.java:
```diff
@@ -20,24 +26,33 @@
             @Parameter(description = "기술면접 ID") @PathVariable(name = "interviewId") Long interviewId,
-            @Parameter(description = "따봉 ID") @PathVariable(name = "thumbId") Long thumbId
+            @LoginUser SessionUser user
     ){
```
따봉 id를 받아오기 어렵기 때문에 기술면접 id로 변경하셨다고 보면 될까요? 놓치고 있던 부분인데 잘 캐치해주셨네요 👍 다만, 하위에 interview가 존재하는 api 구조가 되기 때문에, Restful한 작성법인가에 대한 의문이 들긴 합니다. 기술면접을 조회할 때 thumb 객체(따봉 여부와 따봉 id)를 같이 리턴하거나, api 구조를 손보는 방식으로 가져가면 될 것 같습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/controller/ThumbController.java:
```diff
@@ -48,10 +63,21 @@
+        if (user == null) {
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED);
+        }
     }
```
이왕 글로벌예외 만든 김에 저희 통일하면 어떨까요..? 구조 이해가 어려우시면 제가 해당 부분은 리팩토링 하겠습니다 :)

nrudev replied:
실례가 안된다면 다희님께 리팩토링을 요청드려도 괜찮을까요?

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/controller/ThumbController.java:
```diff
@@ -48,10 +63,21 @@
-        CheckThumbResponse testCheckResponse = new CheckThumbResponse(false, testThumbResponse);
-        return ResponseEntity.ok(testCheckResponse);
+        if (user == null || user.getUserId() != userId) {
+            return new ResponseEntity<>(HttpStatus.NOT_ACCEPTABLE);
```
아마 빌더 패턴으로도 구성할 수 있을텐데, 제가 밖이라... 집에서 다시 코멘트 드리겠습니다!

nrudev replied:
NOT_ACCEPTABLE 에 해당하는 메서드가 없어서 이렇게 처리했습니다! 제가 모르는 걸 수도 있으니 혹시 아는 게 있다면 말씀 부탁드립니다!

dadahee replied:
먼저 해당 부분은 다른 부분과 동일하게 예외 클래스로 관리하는 것이 효율적일 것 같습니다. (+ 예외 메시지도 관리할 수 있으므로)
또한 NOT ACCEPTABLE 에러(406 status code)는 권한 관련 에러가 아니라 유저 처리에는 적합하지 않다고 생각합니다.
그리고 user가 null인 경우와 id가 서로 일치하지 않는 경우(= 타유저 요청) 또한 나누어 각기 다른 예외를 발생시키는 게 올바른 의미 전달이 될 것 같네요. 마지막으로 유저의 권한(관리자/활성화 여부) 또한 체크해야 하기 때문에, 유저가 null인 경우만 컨트롤러에서 처리하고, 나머지는 서비스단에서 처리하는 걸로 하였습니다! 해당 부분은 이견 없으시면 제가 한 번에 리팩토링 하겠습니다.

nrudev replied:
이견 없습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/domain/Thumb.java:
```diff
@@ -14,7 +15,7 @@
 @Getter
 @Entity
-public class Thumb {
+public class Thumb extends BaseTimeEntity {
```
따봉은 수정이 없기 때문에 updated_at 필드가 필요 없지 않을까요?

nrudev replied:
그렇긴 한데 created_at 을 별도로 뒀더니 굳이 null 값을 넣어줘야 하는 귀찮음이 있어서요.
이미 만들어진 BaseTimeEntity를 사용하는 쪽이 낫지 않을까 싶어서 이렇게 했습니다.

nrudev replied:
수정하는 편이 낫다 싶으면 말씀 부탁드립니다!

dadahee replied:
상속받아 사용하는 경우를 공간 효율 관점에서 보면, thumb 하나마다 데이터베이스에 8 bytes(mysql에서 datetime 컬럼의 크기)의 불필요한 공간 차지가 발생하게 됩니다. 개인적으로 해당 문제는 설계상의 문제가 되기 때문에 돌이킬 수 없기 전에(?) 시정하는 편이 좋다고 생각합니다. 번거로우실 줄은 알지만, 수정해주실 수 있을까요? null 관련 부분은 지난 번에 논의했던 `@DynamicInsert` 어노테이션을 이용하면 쉽게 처리하실 수 있을 것 같아요.

nrudev replied:
`@DynamicInsert` 를 사용하니까 null로 들어가네요. 생성자에 `LocalDateTime.now()` 를 넣는 것으로 해결했습니다~

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/repository/ThumbRepository.java:
```diff
@@ -2,6 +2,13 @@
 public interface ThumbRepository extends JpaRepository<Thumb, Long> {
+
+    @Query(nativeQuery = true, value = "SELECT * FROM thumb WHERE interview_id = :interview_id AND user_id = :user_id")
+    Optional<Thumb> findByInterviewAndUser(@Param("interview_id") Long interviewId, @Param("user_id") Long userId);
```
수정할 필요는 없지만, 공부에 참고하시라고 말씀드리자면, findByInterviewIdAndUserId로도 조회 가능해요. !

nrudev replied:
엇 그런가요? 저번에도 말씀드렸듯이 네이밍 전략을 그대로 사용하면 Interview와 User 객체를 넣어줘야 했던 걸로 기억합니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/service/ThumbServiceImpl.java:
```diff
@@ -1,16 +1,52 @@
+  @Transactional
+  @Override
+  public Thumb saveThumb(Long interviewId, Long userId) {
+    User user = userRepository.findById(userId).get();
```
유저도 검증이 필요해보입니다!
컨트롤러에서 널 체크는 하였지만, 실제 존재하는 유저인지 검사도 필요할 것 같아요.

nrudev replied:
말씀이 맞네요! 수정하겠습니다~

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/service/ThumbServiceImpl.java:
```diff
@@ -1,16 +1,52 @@
+  public Thumb saveThumb(Long interviewId, Long userId) {
+    User user = userRepository.findById(userId).get();
+    Interview interview = interviewRepository.findById(interviewId)
+        .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "존재하지 않는 게시물입니다."));
```
말씀드린 것처럼 예외 처리를 통일하면 어떨까요..? 헷갈리시면 마찬가지로 제가 리팩토링 하겠습니다.

nrudev replied:
다희님께 요청드리겠습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/service/ThumbServiceImpl.java:
```diff
@@ -1,16 +1,52 @@
+
+  @Override
+  public boolean checkThumb(Long interviewId, Long userId) {
+    return thumbRepository.findByInterviewAndUser(interviewId, userId).isPresent();
```
깔끔한 코드 정말 좋네요! :)

nrudev replied:
감사합니당:)

dadahee reviewed (commented):
수고 많으셨습니다!

dadahee commented:
고생하셨습니다~ 앞서 언급한 API와 Thumb 엔티티 설계 관련 문제만 해결해주시면 나머지 부분은 제가 변경 및 리팩토링 하겠습니다 💯

nrudev committed:
Merge branch 'develop' of https://github.com/J2KB-5th/chippo into feature/64-back-thumbApi

nrudev committed:
Merge branch 'develop' of https://github.com/J2KB-5th/chippo into feature/64-back-thumbApi

nrudev committed:
chore : 코드리뷰 반영 #83

- Thumb에서 updated_at 삭제
- service에서 user 예외 추가

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/controller/ThumbController.java:
```diff
@@ -1,43 +1,55 @@
 
     @Operation(summary = "따봉 취소", description = "해당 id의 기술면접에서, 해당 id의 따봉(좋아요)을 취소합니다.")
-    @DeleteMapping("/thumbs/{thumbId}")
+    @DeleteMapping("/thumbs")
     public ResponseEntity<Void> deleteThumb(
             @Parameter(description = "기술면접 ID") @PathVariable(name = "interviewId") Long interviewId,
-            @Parameter(description = "따봉 ID") @PathVariable(name = "thumbId") Long thumbId
+            @LoginUser SessionUser user
     ){
```
해당 부분은 변경 없으신 걸까요?

nrudev replied:
사실 더 RESTful 하게 하려면 어떻게 해야 하는지 잘 모르겠어서요ㅠ
다희님이 말씀하신대로 기술면접 조회 부분을 수정하는 쪽이 나을까요..?

dadahee replied:
희승님 아이디어를 방해할까봐 일부러 가만 있었는데, 더 적극적으로 피드백 드릴 걸 그랬네요. 하하
개인적으로는 `DELETE /api/interviews/{interviewId}/users/{userId}/thumb` -> 이 정도로 따봉 데이터는 숨기되, 인터뷰 <-> 유저 <-> 따봉간 관계를 모두 표현할 수 있게 구현하고 유효성 로직을 추가하거나, 기존 URI를 사용하되 기술면접 조회 부분에서 따봉 아이디를 같이 프론트에 전달하는 방식으로 처리하면 어떨까요? 저도 완벽한 건 아니고, 하나의 의견일 뿐이라...ㅎㅎ 이 뒤로는 희승님께 맡기겠습니다 :)

nrudev replied:
말씀해주신대로 수정해봤습니다! 확인 부탁드려여ㅎㅎ

dadahee reviewed (commented):
따봉 취소 부분 피드백 부탁드립니다! 그대로 가져가시기로 한 것 같은데, 이유를 설명해주실 수 있을까요?

nrudev committed:
Merge branch 'develop' of https://github.com/J2KB-5th/chippo into feature/64-back-thumbApi

nrudev committed:
chore : RESTful한 Thumb 취소를 위한 수정(코드리뷰 반영)

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/thumb/controller/ThumbController.java:
```diff
@@ -1,43 +1,56 @@
 
+        checkLogin(user);
+        ThumbResponse response = new ThumbResponse(thumbService.saveThumb(interviewId, user.getUserId()));
         URI uri = uriBuilder.path("/api/interviews/{interviewId}").build().toUri();
-        return ResponseEntity.created(uri).body(testResponse);
+
+        return ResponseEntity.created(uri).body(response);
     }
 
     @Operation(summary = "따봉 취소", description = "해당 id의 기술면접에서, 해당 id의 따봉(좋아요)을 취소합니다.")
-    @DeleteMapping("/thumbs/{thumbId}")
+    @DeleteMapping("/users/{userId}/thumb")
     public ResponseEntity<Void> deleteThumb(
             @Parameter(description = "기술면접 ID") @PathVariable(name = "interviewId") Long interviewId,
-            @Parameter(description = "따봉 ID") @PathVariable(name = "thumbId") Long thumbId
+            @Parameter(description = "사용자 ID") @PathVariable(name = "userId") Long userId,
+            @LoginUser SessionUser user
     ){
+
+        checkLogin(user, userId);
+        thumbService.cancelThumb(interviewId, userId);
+
```
첫번째 방법으로 구현 해주셨네요 :) 수고 많으셨습니다!

dadahee reviewed (commented):
해당 브랜치는 리팩토링 후 병합하겠습니다😄  고생 많으셨어요 💯

Commit:
refactor: Thumb 조회/등록/삭제 API 리팩토링
- 유저 권한 체크 코드 개선: unauthorized / forbidden으로 세분화 및 유효성 검사 위치 변경
- 역할을 명확히 하기 위해 `checkLogin`-> `validateUserAuthentication`으로 메서드명 변경
- 주석 추가

Commit:
refactor: thumb controller 주석 추가
- 각 api의 명령어별 주석 추가
- 반환 시 상태 코드도 함께 명시함

dadahee commented:
리팩토링 후 병합 하였습니다! 확인하신 후에 코멘트 남겨주세요.
(* 브랜치는 확인하시면 삭제 할게요 🙂)
